### PR TITLE
Add support for M1-enabled devices

### PR DESCRIPTION
**Pull request by sywesk**
Hello ! 

This is finally it ! After hundreds of hours of hard work, I can finally submit this merge request that adds all the missing pieces to restore M1 devices.

I know this is not the best code ever, but it works on m1 mac minis. I'd be happy to modify it :)

PS: The code might not work for M1 macbooks, but i'm not a mac user so i won't have any m1 laptop for the foreseeable future.

### CODE REVIEW

File: src/restore.c [hunk 1]
```diff
@@ -932,6 +932,59 @@ int restore_send_filesystem(struct idevicerestore_client_t* client, idevice_t de
 	return 0;
 }
 
+int restore_send_ros_root_ticket(restored_client_t restore, struct idevicerestore_client_t* client)
```
Comment (nikias) [inline src/restore.c:935]: I'd prefer the name `restore_send_recovery_os_root_ticket` here.

File: src/restore.c [hunk 2]
```diff
@@ -2665,6 +2718,398 @@ static int restore_send_firmware_updater_data(restored_client_t restore, struct
...
+
+	info("Done sending %s\n", component_name);
+	return 0;
+}
+
+int restore_send_source_boot_object_v4(restored_client_t restore, struct idevicerestore_client_t* client, plist_t msg, plist_t build_identity) {
+	debug_plist(msg);
+
+	char *image_name = NULL;
+	// Get checkpoint id
```
Comment (nikias) [inline src/restore.c:2933]: I think this comment is wrong

File: src/restore.c [hunk 3]
```diff
@@ -2665,6 +2718,398 @@ static int restore_send_firmware_updater_data(restored_client_t restore, struct
...
+	free(ticket_path);
+
+	return 0;
+}
+
+int restore_send_personalized_boot_object_v3(restored_client_t restore, struct idevicerestore_client_t* client, plist_t msg, plist_t build_identity) {
+	debug_plist(msg);
+
+	char *image_name = NULL;
+	// Get checkpoint id
```
Comment (nikias) [inline src/restore.c:2813]: I think this comment is wrong.

File: src/restore.c [hunk 4]
```diff
@@ -2665,6 +2718,398 @@ static int restore_send_firmware_updater_data(restored_client_t restore, struct
...
+	plist_t args = plist_dict_get_item(msg, "Arguments");
+	plist_t is_recovery_node = plist_dict_get_item(args, "IsRecoveryOS");
+	uint8_t is_recovery = 0;
+	plist_get_bool_val(is_recovery_node, &is_recovery);
+
+	return restore_get_build_identity(client, is_recovery);
+}
+
+int extract_macos_variant(plist_t build_identity, char** output) {
+	// Get the device class from the build identity. `image_name` should look like 'j293ap' at the end.
```
Comment (nikias) [inline src/restore.c:2751]: Drop this comment, I guess.

File: src/restore.c (hunk 2 above)
Comment (sywesk) [reply src/restore.c:2933]: done :heavy_check_mark:

File: src/restore.c (hunk 4 above)
Comment (sywesk) [reply src/restore.c:2751]: done :heavy_check_mark:

**Comment by sywesk** [comment]
I did address all of your comments, I'll let you resolve the conversations if you're happy with the changes :+1:

**Comment by marcan** [comment]
Awesome work! I'm happy to test this on an M1 Macbook Pro and Macbook Air if you want :) (at least for SFR/1TR restore, not sure I want to reinstall everything right now :-))

**Comment by sywesk** [comment]
@marcan I'm not sure about what you mean, but I think you're talking about doing an "Update" (restore behavior) instead of an "Erase", so that only the macOS system is rewritten ? If yes, I haven't tested this behavior. But if you want to help test macbooks, i'll first make sure it works for mac minis :smile: Just to try to minimize the risk :pray: 

(Btw, thanks for the awesome macvdmtool and more generally your work on asahilinux, it helped us a lot in the beginning :heart: )

**Comment by rnagy** [comment]
Hi

I am trying to restore my M1 MacbookPro and it fails with all firmwares:

`

Found device in Recovery mode
Identified device as j293ap, MacBookPro17,1
The following firmwares are currently being signed for MacBookPro17,1:
  [1] 11.4 (build 20F71)
  [2] 11.3.1 (build 20E241)
  [3] 11.3 (build 20E232)
  [4] 11.2.3 (build 20D91)
  [5] 11.2.2 (build 20D80)
  [6] 11.2.1 (build 20D74)
  [7] 11.2 (build 20D64)
  [8] 11.1 (build 20C69)
  [9] 11.0.1 (build 20B29)
Select the firmware you want to restore: 3
Selected firmware 11.3 (build 20E232)
Verifying 'UniversalMac_11.3_20E232_Restore.ipsw'...
Extracting BuildManifest from IPSW
Product Version: 11.3
Product Build: 20E232 Major: 20
INFO: device serial number is FVFFK08DQ05P
Device supports Image4: true
Variant: Customer Erase Install (IPSW)
This restore will erase your device data.
################################ [ WARNING ] #################################
# You are about to perform an *ERASE* restore. ALL DATA on the target device #
# will be IRREVERSIBLY DESTROYED. If you want to update your device without  #
# erasing the user data, hit CTRL+C now and restart without -e or --erase    #
# command line switch.                                                       #
# If you want to continue with the ERASE, please type YES and press ENTER.   #
##############################################################################
> YES
Checking IPSW for required components...
All required components found in IPSW
Using cached filesystem from 'UniversalMac_11.3_20E232_Restore/038-69368-243.dmg'
Found ECID 2375052739346462
Getting ApNonce in recovery mode... 48 99 82 54 b2 c7 57 bd b9 8e 65 f8 54 79 aa 7a 87 08 b0 c8 fe ef e8 0d 8b 66 b9 aa 1d 52 40 eb 
Trying to fetch new SHSH blob
Getting SepNonce in recovery mode... f2 72 18 d4 ca 1b b8 e9 f6 02 17 6f 7d a1 b9 e2 71 73 1b ef 
Request URL set to https://gs.apple.com/TSS/controller?action=2
Sending TSS request attempt 1... response successfully received
Received SHSH blobs
Getting SepNonce in recovery mode... f2 72 18 d4 ca 1b b8 e9 f6 02 17 6f 7d a1 b9 e2 71 73 1b ef 
Request URL set to https://gs.apple.com/TSS/controller?action=2
Sending TSS request attempt 1... response successfully received
Received SHSH blobs
Getting SepNonce in recovery mode... f2 72 18 d4 ca 1b b8 e9 f6 02 17 6f 7d a1 b9 e2 71 73 1b ef 
Request URL set to https://gs.apple.com/TSS/controller?action=2
TSS server returned: STATUS=8&MESSAGE=An internal error occurred.
ERROR: TSS request failed (status=8, message=An internal error occurred.)
Sending TSS request attempt 1... ERROR: Unable to send TSS request
ERROR: Unable to get SHSH blobs for this device (recovery OS Root Ticket)
`

**Comment by nikias** [comment]
@rnagy can you open a ticket for this? Please run with -d since it seems like TSS doesn't like the request.

**Comment by rnagy** [comment]
> @rnagy can you open a ticket for this? Please run with -d since it seems like TSS doesn't like the request.

#411 has been created

**Pull request merged**